Main svc

Svetlana-Seregina opened the pull request.

Svetlana-Seregina committed:
added methods for users in AdminController.java

Svetlana-Seregina committed:
added methods for category

Svetlana-Seregina committed:
StatsClient and MainService

Svetlana-Seregina committed:
all tests passed

Svetlana-Seregina committed:
all tests passed

Svetlana-Seregina committed:
Part 2, review 1

Svetlana-Seregina committed:
Part 2, review 1

Svetlana-Seregina committed:
Part 2, Review 1

Svetlana-Seregina committed:
Part 2, review 1

Svetlana-Seregina committed:
Part 2, review 1

Svetlana-Seregina committed:
Part 2, review 1

Svetlana-Seregina committed:
Part 2, review 1

Svetlana-Seregina committed:
Part 2, review 1

Svetlana-Seregina committed:
Part 2, review 1

TyutterinYakov commented on main-service/pom.xml:
```diff
@@ -69,6 +69,29 @@
+            <artifactId>spring-tx</artifactId>
+        </dependency>
+        <!--<dependency>
+            <groupId>ru.practicum</groupId>
```
Если зависимость не требуется, следует ее исключить

TyutterinYakov commented on main-service/pom.xml:
```diff
@@ -69,6 +69,29 @@
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-tx</artifactId>
```
Для чего нам две такие зависимости?)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/Location.java:
```diff
@@ -0,0 +1,25 @@
+@ToString
+@AllArgsConstructor
+@NoArgsConstructor
+public class Location {
```
(предложение) Если будет интересно, то для локации было бы удобно использовать составной первичный ключ, который будет представлен `lat` и `lon`
https://www.baeldung.com/jpa-composite-primary-keys

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/repository/ParticipationRequestRepository.java:
```diff
@@ -0,0 +1,24 @@
+
+public interface ParticipationRequestRepository extends JpaRepository<ParticipationRequest, Long> {
+
+    Optional<ParticipationRequest> findByRequester(UserDto userDto);
```
Действительно ли это корректный метод? Разве к одному отправителю может относиться только один запрос на участие?) У него может быть их несколько для разных событий

TyutterinYakov commented on stats-service/stats-server/src/main/java/ru/practicum/explorewithme/EndpointHitRepository.java:
```diff
@@ -9,7 +9,7 @@
 
     @Query(value = "SELECT " +
-            "distinct new ru.practicum.explorewithme.ViewStats(a.name, ep.uri, COUNT(ep.uri)) " +
+            "new ru.practicum.explorewithme.ViewStats(a.name, ep.uri, COUNT(distinct ep.uri)) " +
```
`distinct ep.ip`*, так как уникальность определяется уникальностью ip-адреса, с которого пришел запрос)

TyutterinYakov commented on stats-service/stats-server/src/main/java/ru/practicum/explorewithme/EndpointHitRepository.java:
```diff
@@ -18,13 +18,13 @@ public interface EndpointHitRepository extends JpaRepository<EndpointHit, Long>
 
     @Query(value = "SELECT " +
-            "distinct new ru.practicum.explorewithme.ViewStats(a.name, ep.uri, COUNT(ep.uri)) " +
+            "new ru.practicum.explorewithme.ViewStats(a.name, ep.uri, COUNT(distinct ep.uri)) " +
```
`distinct ep.ip`*, так как уникальность определяется уникальностью ip-адреса, с которого пришел запрос)

TyutterinYakov commented on stats-service/stats-server/src/main/java/ru/practicum/explorewithme/EndpointHitRepository.java:
```diff
@@ -18,13 +18,13 @@ public interface EndpointHitRepository extends JpaRepository<EndpointHit, Long>
             "AND ep.uri IN ?3 " +
             "GROUP BY a.name, ep.uri " +
-            "ORDER BY count(ep.uri) desc ")
+            "ORDER BY count(distinct ep.uri) desc ")
```
distinct ep.ip*

TyutterinYakov commented on stats-service/stats-server/src/main/java/ru/practicum/explorewithme/EndpointHitRepository.java:
```diff
@@ -18,13 +18,13 @@ public interface EndpointHitRepository extends JpaRepository<EndpointHit, Long>
     List<ViewStats> findViewStatsWhenUrisIsEmptyAndIpIsUnique(LocalDateTime start, LocalDateTime end);
```
distinct ep.ip*

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/AdminController.java:
```diff
@@ -0,0 +1,125 @@
+@RequiredArgsConstructor
+@Validated
+@Slf4j
+public class AdminController {
```
Следует делить все контроллеры не только по принадлежности к API администратора/публичного/пользовательского, но и по работе с определенными сущностями. То есть у нас должен быть, например, контроллер публичного/приватного/административного API для событий и тд

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/PrivateController.java:
```diff
@@ -0,0 +1,107 @@
+@RequiredArgsConstructor
+@Validated
+@Slf4j
+public class PrivateController {
```
Следует делить все контроллеры не только по принадлежности к API администратора/публичного/пользовательского, но и по работе с определенными сущностями. То есть у нас должен быть, например, контроллер публичного/приватного/административного API для событий и тд

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/PublicController.java:
```diff
@@ -0,0 +1,112 @@
+@RestController
+@RequiredArgsConstructor
+@Slf4j
+public class PublicController {
```
Следует делить все контроллеры не только по принадлежности к API администратора/публичного/пользовательского, но и по работе с определенными сущностями. То есть у нас должен быть, например, контроллер публичного/приватного/административного API для событий и тд

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/category/CategoryDto.java:
```diff
@@ -0,0 +1,26 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+public class CategoryDto {
```
Разве это DTO?) Это класс-сущность
Объект Передачи Данных (DTO) — шаблон проектирования, используемый для передачи данных между программными системами или уровнями в программной архитектуре. Это простой объект содержащий набор данных, обычно в форме свойств или полей, которые представляют определённый объект или концепцию в системе.
Приставку DTO следует исключить. Также проверь, пожалуйста, другие классы) Также в качестве возвращаемых/принимаемых значений в методах контроллеров должны выступать именно DTO

TyutterinYakov replied:
Для соблюдения инкапсуляции - поля класса должны быть приватными

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/user/UserDto.java:
```diff
@@ -0,0 +1,23 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+public class UserDto {
```
Разве это DTO?) Это класс-сущность
Объект Передачи Данных (DTO) — шаблон проектирования, используемый для передачи данных между программными системами или уровнями в программной архитектуре. Это простой объект содержащий набор данных, обычно в форме свойств или полей, которые представляют определённый объект или концепцию в системе.
Приставку DTO следует исключить. Также проверь, пожалуйста, другие классы) Также в качестве возвращаемых/принимаемых значений в методах контроллеров должны выступать именно DTO

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/category/NewCategoryDto.java:
```diff
@@ -0,0 +1,19 @@
+public class NewCategoryDto {
+
+    @NotBlank
+    @Column(unique = true)
```
Действительно ли здесь требуется эта аннотация?)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/category/NewCategoryDto.java:
```diff
@@ -0,0 +1,19 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class NewCategoryDto {
```
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/PublicController.java:
```diff
@@ -0,0 +1,112 @@
+        log.info("client ip: {}", ip);
+        String path = request.getRequestURI();
+        log.info("endpoint path: {}", path);
+        String app = "ewm-main-service";
```
Более корректно было бы для названия приложения - app ввести константу, но ее значение мы будем хранить не в коде, так как, если потребуется запустить несколько экземпляров приложения и требованием будет уникальное имя для каждого, каждый раз лезть в код, чтобы его изменить, будет не совсем логично. Его можно будет разместить в файле конфигурации `application.properties`, а получение производить с помощью `@Value`, также потом, при необходимости, его можно будет передать через переменные среды)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/PublicController.java:
```diff
@@ -0,0 +1,112 @@
+                text, categories, paid, rangeStart, rangeEnd, onlyAvailable, sort, from, size);
+
+        String path = request.getRequestURI();
+        String app = "ewm-main-service";
```
Более корректно было бы для названия приложения - app ввести константу, но ее значение мы будем хранить не в коде, так как, если потребуется запустить несколько экземпляров приложения и требованием будет уникальное имя для каждого, каждый раз лезть в код, чтобы его изменить, будет не совсем логично. Его можно будет разместить в файле конфигурации `application.properties`, а получение производить с помощью `@Value`, также потом, при необходимости, его можно будет передать через переменные среды)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/PublicController.java:
```diff
@@ -0,0 +1,112 @@
+
+        EndpointHitDto endpointHitDto = new EndpointHitDto(app, path, ip, LocalDateTime.now());
+        log.info("Передаем endpointHitDto в statsClient: {}", endpointHitDto);
+        statsClient.save(endpointHitDto);
```
Хорошо было бы преобразовать метод таким образом, чтобы он принимал `HttpServletRequest`, чтобы вся остальная дублирующаяся логика находилась внутри метода клиента)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/PublicController.java:
```diff
@@ -0,0 +1,112 @@
+
+        EndpointHitDto endpointHitDto = new EndpointHitDto(app, path, ip, LocalDateTime.now());
+        log.info("Передаем endpointHitDto в statsClient: {}", endpointHitDto);
+        statsClient.save(endpointHitDto);
```
Хорошо было бы преобразовать метод таким образом, чтобы он принимал `HttpServletRequest`, чтобы вся остальная дублирующаяся логика находилась внутри метода клиента)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/compilation/NewCompilationDto.java:
```diff
@@ -0,0 +1,18 @@
+
+    private final List<Long> events;
+
+    private final Boolean pinned;
```
`pinned` может иметь лишь два состояния, хорошо было бы представить его в виде примитива, чтобы, если значение не будет передано - оно было проинициализировано как `false`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/compilation/NewCompilationDto.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+
+@Data
+public class NewCompilationDto {
```
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/compilation/UpdateCompilationRequest.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.List;
+
+@Data
+public class UpdateCompilationRequest {
```
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`. Эта аннотация удобна тем, что допускает null, за счет чего при обновлении по значению мы сможем понять, было передано значение или нет)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/Event.java:
```diff
@@ -0,0 +1,68 @@
+
+    private String title; // Заголовок
+
+    private Long views; // Количество просмотрев события
```
Хранение вычисляемых полей в БД - не является хорошей практикой. Хорошо было бы исключить это поле из таблицы БД
Также в ТЗ явно указано, что хранение просмотров должно осуществляться в сервисе статистики)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/Event.java:
```diff
@@ -0,0 +1,68 @@
+    private CategoryDto category;
+
+    @Column(name = "confirmed_requests")
+    private Long confirmedRequests; // Количество одобренных заявок на участие в данном событии
```
Хранение вычисляемых полей в БД - не является хорошей практикой. Хорошо было бы исключить это поле из таблицы БД

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventAdminRequest.java:
```diff
@@ -0,0 +1,41 @@
+
+    private Boolean paid;
+
+    private Long participantLimit;
```
`participantLimit` может быть только положительным число, либо же 0. В данном случае было бы удобно использовать `@PositiveOrZero`, так как она удобна тем, что допускает null, по которому при обновлении мы сможем понять, было передано значение или нет

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventAdminRequest.java:
```diff
@@ -0,0 +1,41 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate;
+
+    private Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventUserRequest.java:
```diff
@@ -0,0 +1,38 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private final LocalDateTime eventDate;
+
+    private final Location location;
```
Классы-DTO не должны содержать в себе сложные объекты по типу классов-сущностей. Мы всегда должны контролировать те данные, что получаем и отдаем. Не только в целях безопасности, но и производительности, так как зачем тратить ресурсы на передачу тех данных, что не будут использоваться

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,47 @@
+
+    private Boolean paid; // default: false; Нужно ли оплачивать участие в событии
+
+    private Long participantLimit;
```
`participantLimit` может быть только положительным число, либо же 0. В данном случае было бы удобно использовать `@PositiveOrZero`
Также хорошо было бы представить его в виде примитива, чтобы, если значение не будет задано, оно было проинициализировано нулем)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,47 @@
+
+    private Location location;
+
+    private Boolean paid; // default: false; Нужно ли оплачивать участие в событии
```
`paid` может иметь лишь два состояния, хорошо было бы представить его в виде примитива)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,47 @@
+
+    private Long participantLimit;
+
+    private Boolean requestModeration; // Нужна ли пре-модерация заявок на участие.
```
`requestModeration` может иметь лишь два состояния, хорошо было бы представить его в виде примитива)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,47 @@
+    @Size(min = 20, max = 2000)
+    private String annotation;
+
+    private Long category;
```
События без категории быть не может) Следует навесить над полем `@NotNull`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,47 @@
+    // Если true, то все заявки будут ожидать подтверждения инициатором события.
+    // Если false - то будут подтверждаться автоматически.
+
+    @Size(min = 3, max = 120)
```
Не забываем также использовать `@NotBlank`, так как строка не должна состоять только из пробелов и значение не должно быть null

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventUserRequest.java:
```diff
@@ -0,0 +1,38 @@
+
+    private final Boolean requestModeration;
+
+    private final String stateAction; // Изменение сотояния события: SEND_TO_REVIEW, CANCEL_REVIEW
```
Можно было бы ввести для `stateAction` Enum внутри этого класса, чтобы ограничить выбор)
```
	public enum StateAction {
		SEND_TO_REVIEW,
		CANCEL_REVIEW
	}
```
Также можно было бы ввести общего родителя для этого класса и класса DTO обновления события администратором, так как их единственным отличие будет это поле, которое как раз и будет вынесено в наследников)
У администратора же будет 
```
	public enum StateAction {
		PUBLISH_EVENT,
		REJECT_EVENT
	}
```

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventUserRequest.java:
```diff
@@ -0,0 +1,38 @@
+    @Size(min = 20, max = 2000)
+    private final String annotation;
+
+    private final CategoryDto category;
```
Для чего мы представляем его в виде DTO, если в спецификации прописано, что это `Long`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/AdminServiceImpl.java:
Аналогично и сервисы следует разбить на отдельные, но там деление на административный/приватный/публичный API необязательно)

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего вечера! Если возникнут вопросы - пиши в пачку, буду рад помочь!🌃

Svetlana-Seregina committed:
pom.main-service; ParticipationRequestRepository change method to findByRequesterAndEvent;
created ValidationException; updateEventById in AdminService;
UpdateEventUserRequest added @NoArgsConstructor @ToString;
updateEventByInitiator added if and event;
distinct ep.ip in EndpointHitrepository;
add Category; add User;

Svetlana-Seregina committed:
refactoring controllers and servers; variable app in properties;
constraint annotations

Svetlana-Seregina committed:
Part 2, review 2

Svetlana-Seregina committed:
Part 2, review 2

Svetlana-Seregina committed:
Part 2, review 2

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/category/CategoryAdminController.java:
```diff
@@ -0,0 +1,44 @@
+    public ResponseEntity<Boolean> deleteCategoryById(@PathVariable long catId) {
+        log.info("Обрабатываем запрос на удаление категории. Создан запрос на удаление категории по id = {}", catId);
+        boolean categoryDelete = categoryService.deleteCategoryById(catId);
+        return new ResponseEntity<>(categoryDelete, HttpStatus.NO_CONTENT);
```
Передача тела ответа не имеет смысла, так как при использовании `NO_CONTENT` 204, оно отсутствует

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/category/CategoryPublicController.java:
```diff
@@ -0,0 +1,40 @@
+        try {
+            CategoryDto category = categoryService.findCategoryById(catId);
+            return new ResponseEntity<>(category, HttpStatus.OK);
+        } catch (EntityNotFoundException e) {
```
Почему бы не добавить отдельный хендлер для обработки этого исключения для возврата требуемого кода ответа, чтобы не производить обработку в коде?)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/category/CategoryPublicController.java:
```diff
@@ -0,0 +1,40 @@
+    private final CategoryService categoryService;
+
+    @GetMapping
+    public List<CategoryDto> findAllCategories(@RequestParam(value = "from", defaultValue = "0") Integer from,
```
(предложение)Почему бы не использовать более декларативный подход и использовать аннотации ограничения) Например, для from можно было бы использовать `@PositiveOrZero`, так как значение может быть 0, либо больше) Аналогично и для size, но в этом случае следует использовать `@Positive`, так как значение не может быть равно 0
Кстати, для того, чтобы эти аннотации стали учитываться, необходимо над классом навесить аннотацию `@Validated`. Но выбрасываться будет не `MethodArgumentNotValidException`, к которому мы привыкли, а `ConstraintViolationException`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/compilation/CompilationAdminController.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @PatchMapping("/{compId}")
+    public CompilationDto updateCompilationById(@RequestBody UpdateCompilationRequest updateCompilationRequest,
```
Перед аргументом `UpdateCompilationRequest` отсутствует `@Valid`, из-за чего аннотации валидации из него не будут учтены)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/category/CategoryDto.java:
```diff
@@ -0,0 +1,14 @@
+import javax.validation.constraints.NotBlank;
+
+@Data
+public class CategoryDto {
```
Хорошо было бы ограничить размеры текстовых полей на основе тех ограничений, которые установлены на уровне таблицы в БД, чтобы заранее откинуть некорректные запросы) Для ограничения количества символов можно использовать `@Size`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/category/Category.java:
```diff
@@ -0,0 +1,25 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+public class Category {
```
Для соблюдения инкапсуляции - поля класса должны быть приватными)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/user/UserAdminController.java:
```diff
@@ -0,0 +1,46 @@
+        log.info("Обрабатываем запрос на удаление пользователя с id = {}", userId);
+        boolean deleteUser = userService.deleteUserById(userId);
+        log.info("Пользователь удален ? = {}", deleteUser);
+        return new ResponseEntity<>(deleteUser, HttpStatus.NO_CONTENT);
```
Аналогично, передача тела ответа при возврате статуса 204 - не имеет смысла, так как у него отсутствует тело ответа

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/compilation/CompilationAdminController.java:
```diff
@@ -0,0 +1,44 @@
+    public ResponseEntity<Boolean> deleteCompilation(@PathVariable long compId) {
+        log.info("Обрабатываем запрос на удаление подборки, id = {}", compId);
+        boolean deleteCompilation = compilationService.deleteCompilation(compId);
+        return new ResponseEntity<>(deleteCompilation, HttpStatus.NO_CONTENT);
```
Аналогично, возврат значения в теле ответа не имеет смысла

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/event/EventPublicController.java:
```diff
@@ -0,0 +1,75 @@
+    private final StatsClient statsClient;
+
+    @Value("${app}")
+    String app;
```
Для соблюдения инкапсуляции - поля класса должны быть приватными)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/event/EventPublicController.java:
```diff
@@ -0,0 +1,75 @@
+                                             @RequestParam(value = "paid", required = false) Boolean paid,
+                                             @RequestParam(value = "rangeStart", required = false) @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") LocalDateTime rangeStart,
+                                             @RequestParam(value = "rangeEnd", required = false) @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") LocalDateTime rangeEnd,
+                                             @RequestParam(value = "onlyAvailable", required = false) Boolean onlyAvailable,
```
`onlyAvailable` по спецификации является необязательным, но для него должно быть указано стандартное значение `defaultValue = "false"`, `required = false` следует исключить)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/event/EventPublicController.java:
```diff
@@ -0,0 +1,75 @@
+                                             @RequestParam(value = "rangeStart", required = false) @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") LocalDateTime rangeStart,
+                                             @RequestParam(value = "rangeEnd", required = false) @DateTimeFormat(pattern = "yyyy-MM-dd HH:mm:ss") LocalDateTime rangeEnd,
+                                             @RequestParam(value = "onlyAvailable", required = false) Boolean onlyAvailable,
+                                             @RequestParam(value = "sort", required = false) String sort, // EVENT_DATE, VIEWS
```
Тип сортировки `sort` по спецификации является необязательным, но для него должно быть указано стандартное значение `defaultValue = "EVENT_DATE"`, `required = false` следует исключить)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/event/EventPublicController.java:
```diff
@@ -0,0 +1,75 @@
+        String path = request.getRequestURI();
+        String ip = request.getRemoteAddr();
+
+        EndpointHitDto endpointHitDto = new EndpointHitDto(app, path, ip, LocalDateTime.now());
```
Хорошо было бы создать отдельный сервис, в который будет заижекчен `StatsClient`. Таким образом, в этом сервисе будет метод, принимающий на вход `HttpServletRequest`, чтобы не дублировать логику) Также в полях этого сервиса будет `String app;`, так как название приложения будет задаваться на его уровне

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventRequest.java:
```diff
@@ -0,0 +1,40 @@
+
+    protected Boolean requestModeration;
+
+    protected String stateAction;
```
Поле не требуется, так как оно присутствует в классах-наследниках, с указанием нужного типа)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventAdminRequest.java:
Ну просто сказка)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/UpdateEventUserRequest.java:
Ну просто сказка)

TyutterinYakov commented on main-service/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,73 @@
+    CONSTRAINT fk_event_id FOREIGN KEY (event_id) REFERENCES events (id)
+);
+
+CREATE TABLE IF NOT EXISTS list_of_entities
```
Первичный ключ в связующей таблице должен выступать в виде отдельного столбца, либо же быть составным

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/category/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,100 @@
+    public boolean deleteCategoryById(long catId) {
+        Category category = categoryRepository.findById(catId)
+                .orElseThrow(() -> new EntityNotFoundException(String.format("Категории с id = %d нет в базе.", catId)));
+        List<Event> events = eventRepository.findAllByCategory(category);
```
(предложение) Можно было бы использовать не `findAll...`, а `exists...`, чтобы нам вернулось просто `boolean` значение, на основе которого можно будет понять, присутствуют события, связанные с категорией или нет) Таким образом мы не будем тратить ресурсы на десериализацию данных

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/category/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,100 @@
+
+        if (categoryDto.getName() != null && !categoryDto.getName().isBlank()) {
+            String name = categoryDto.getName();
+            if (categoryRepository.findByName(name).isPresent()) {
```
А если на обновление придет категория с именем, равным тому, что было до обновления?) Корректно ли отработает эта логика?

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/request/RequestServiceImpl.java:
```diff
@@ -0,0 +1,241 @@
+        log.info("event.getConfirmedRequests() = {}", confirmedRequests);
+        log.info("event.getParticipantLimit() = {}", event.getParticipantLimit());
+
+        if (confirmedRequests >= event.getParticipantLimit()) {
```
Для начала следует проверить, что `event.getParticipantLimit() > 0`, так как, если значение равно 0, значит ограничения отсутствуют)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,52 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate;
+
+    private Location location;
```
Локация не должна отсутствовать) Следует навесить над полем аннотацию `@NotNull`
Также можно было бы в классе `Location` использовать аннотации валидации, а над полем навесить `@Valid`, чтобы они стали учитываться)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+    @Override
+    public CompilationDto createNewCompilation(NewCompilationDto newCompilationDto) {
+        List<Long> eventIds = newCompilationDto.getEvents();
+        if (eventIds.isEmpty()) {
```
Для начала следует проверить, что значение не `null`, иначе есть шанс возникновения NPE

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+        List<Long> eventIds = updateCompilationRequest.getEvents();
+        log.info("Количество eventIds: {}", eventIds);
+
+        if (eventIds.isEmpty()) {
```
Для начала следует проверить, что значение не `null`, иначе есть шанс возникновения NPE

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+        if (eventIds.isEmpty()) {
+            Set<Event> eventSet = new HashSet<>();
+            compilation.setEvents(eventSet);
+            compilation.setPinned(updateCompilationRequest.getPinned());
```
Отсутствует какая-либо проверки значений `pinned` и `title` перед установкой, так как они не являются обязательными)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,147 @@
+        Set<Event> eventSet = new HashSet<>(events);
+
+        compilation.setEvents(eventSet);
+        compilation.setPinned(updateCompilationRequest.getPinned());
```
Отсутствует какая-либо проверки значений `pinned` и `title` перед установкой, так как они не являются обязательными)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,492 @@
+        Map<Long, List<ParticipationRequest>> confirmedRequestsByEventId =
+                confirmedRequests
+                        .stream()
+                        .collect(Collectors.groupingBy(b -> b.getEvent().getId(), toList()));
```
Можно было бы использовать `counting()`, вместо `toList()`, чтобы сразу получить мапу в виде `Map<Long, Long>`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,492 @@
+    public EventFullDto findEventById(long id, String path) {
+        Event event = findEventInRepository(id);
+
+        List<ParticipationRequest> participationRequests = participationRequestRepository.findAllByEventIdAndStatusIs(
```
Почему бы вместо `findAll...` не использовать `count....`, это было бы более оптимально) Так как мы бы сразу получили количество записей и тратили бы ресурсы на десериализацию, которая требуется лишь для подсчета количества

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,492 @@
+
+        List<String> uris = new ArrayList<>();
+        uris.add(path);
+        List<ViewStats> viewStats = statsClient.getStats(publishedDate, actualDate, uris, false);
```
Получать мы должны уникальные просмотры)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,492 @@
+        }
+        log.info("СПИСОК URIS = {}", uris);
+
+        Map<Long, List<ViewStats>> views = statsClient.getStats(publishedDate, actualDate, uris, false)
```
Получать мы должны уникальные просмотры)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,492 @@
+
+        Map<Long, List<ViewStats>> views = statsClient.getStats(publishedDate, actualDate, uris, false)
+                .stream()
+                .collect(Collectors.groupingBy(b -> (long) Integer.parseInt(b.getUri().substring(8)), toList()));
```
Вместо `gropingBy` хорошо было бы использовать `toMap`, так как мы сможем сразу же сформировать мапу вида: `Map<Long, Long>`, где ключ - идентификатор события, а значение - количество просмотров)

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌞

Svetlana-Seregina committed:
deleteCategoryById(); findCategoryById() -> EntityNotFoundException.class ErrorHandler;
updateCompilationById() @Valid; CategoryDto - @Size; Category private;
deleteUserById(); deleteCompilation(); UpdateEventRequest - stateAction;

Svetlana-Seregina committed:
commit 5

Svetlana-Seregina committed:
Part 2, review 3

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/Location.java:
```diff
@@ -0,0 +1,31 @@
+    private Double lat; // широта места проведения события
+
+    @NotNull
+    @Valid
```
Для чего здесь мы используем `@Valid`?) В классе `Double` отсутствуют аннотации валидации)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/Location.java:
```diff
@@ -0,0 +1,31 @@
+    private Long id;
+
+    @NotNull
+    @Valid
```
Для чего здесь мы используем `@Valid`?) В классе `Double` отсутствуют аннотации валидации)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Data
+    public static class Location {
+        private final double lat;
```
Я имел в виду использовать запись вида:
```
@NotNull
private final Double lat;
@NotNull
private final Double lon;
```

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,53 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate;
+
+    @NotNull
```
А здесь мы навесим `@Valid`, чтобы аннотации валидации из внутреннего класса стали учитываться)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/controller/user/UserAdminController.java:
```diff
@@ -0,0 +1,46 @@
+    private final UserService userService;
+
+    @GetMapping
+    public List<UserDto> findAllUsers(@RequestParam(value = "ids") List<Long> ids,
```
Параметр `ids` не является обязательным) Следует это указать с помощью `required = false` и учитывать в логике

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/request/RequestServiceImpl.java:
```diff
@@ -0,0 +1,244 @@
+
+        log.info("Размер списка participationRequestList = {}", participationRequestList.size());
+        if (eventRequestStatusUpdateRequest.getStatus().equals("CONFIRMED")) {
+            if (event.getParticipantLimit() <= participationRequestList.size()) {
```
Для начала следует проверить, что `participantLimit != 0`, так как, если значение равно 0 - значит ограничения отсутствуют)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+        compilation.setEvents(eventSet);
+        compilation.setPinned(updateCompilationRequest.getPinned() != null ?
+                updateCompilationRequest.getPinned() : compilation.getPinned());
+        compilation.setTitle(updateCompilationRequest.getTitle() != null ?
```
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода `isBlank`)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+            compilation.setEvents(eventSet);
+            compilation.setPinned(updateCompilationRequest.getPinned() != null ?
+                    updateCompilationRequest.getPinned() : compilation.getPinned());
+            compilation.setTitle(updateCompilationRequest.getTitle() != null ?
```
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода `isBlank`)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/compilation/CompilationServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+
+        Set<Event> eventSet = new HashSet<>(events);
+
+        compilation.setEvents(eventSet);
```
Хорошо было бы всю логику по установке событий подборки вынести в условие выше, а всю остальную логику держать вне этого условия) Таким образом, мы избавимся от дублирования кода с установкой заголовка, pinned  и тд)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,481 @@
+
+        Map<Long, Long> views = statsClient.getStats(publishedDate, actualDate, uris, true)
+                .stream()
+                .collect(Collectors.groupingBy(b -> (long) Integer.parseInt(b.getUri().substring(8)), Collectors.counting()));
```
А для чего мы здесь используем `groupingBy`?) Нам возвращается коллекция `ViewStats`, где прописан url, к которому относится количество просмотров конкретно для этого url, то есть повторы исключены, так как на уровне запроса к БД в модуле статистики мы производим группировку по url
Вместо него следует использовать `toMap`

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,481 @@
+                    if (confirmedRequestsByEventId.isEmpty()) {
+                        return eventFullDto;
+                    }
+                    Long confirmedRequests = confirmedRequestsByEventId.get(eventFullDto.getId());
```
В данном случае хорошо было бы использовать `getOrDefault`, чтобы, если значение не будет найдено в мапе, можно было бы задать дефолтное - 0L

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,481 @@
+                    if (views.isEmpty()) {
+                        return eventFullDto;
+                    }
+                    Long allViews = views.get(eventFullDto.getId());
```
В данном случае хорошо было бы использовать `getOrDefault`, чтобы, если значение не будет найдено в мапе, можно было бы задать дефолтное - 0L

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,481 @@
+                    if (confirmedRequestsByEventId.isEmpty()) {
+                        return eventShortDto;
+                    }
+                    Long confirmedRequests = confirmedRequestsByEventId.get(eventShortDto.getId());
```
В данном случае хорошо было бы использовать `getOrDefault`, чтобы, если значение не будет найдено в мапе, можно было бы задать дефолтное - 0L

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/service/event/EventServiceImpl.java:
```diff
@@ -0,0 +1,481 @@
+                    if (views.isEmpty()) {
+                        return eventShortDto;
+                    }
+                    Long allViews = views.get(eventShortDto.getId());
```
В данном случае хорошо было бы использовать `getOrDefault`, чтобы, если значение не будет найдено в мапе, можно было бы задать дефолтное - 0L

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌅

Svetlana-Seregina committed:
Part 2, review 4

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,56 @@
+    @Data
+    public static class Location {
+        @NotNull
+        private final double lat;
```
В качестве типа должен выступать ссылочный - `Double`, так как `@NotNull` не используется над примитивами)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,56 @@
+        @NotNull
+        private final double lat;
+        @NotNull
+        private final double lon;
```
В качестве типа должен выступать ссылочный - `Double`, так как `@NotNull` не используется над примитивами)

TyutterinYakov commented on main-service/src/main/java/ru/practicum/explorewithme/dto/event/NewEventDto.java:
```diff
@@ -0,0 +1,56 @@
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss")
+    private LocalDateTime eventDate;
+
+    @Valid
```
Также следует навесить над полем `@NotNull`, так как иначе возможно вообще не передавать это значение)

TyutterinYakov reviewed (approved):
Добрый день! Ну что ж, поздравляю с успешной сдачей спринта) Отметил еще парочку недочетов, исправь их, пожалуйста, к фиче) Хорошего дня и отличного настроения!☀️🎇

Svetlana-Seregina committed:
@NotNull Location
